### PR TITLE
fix(sync): up to date stagingSites

### PR DESCRIPTION
**Pull request by nightnei**
## Related issues

- Fixes https://github.com/Automattic/dotcom-forge/issues/9975

## Proposed Changes
With this PR we are keeping appdata-v1.json in up-to-date state for `name` and `stagingSites`.

## Testing Instructions
Checkout to this branch and run `STUDIO_SITE_SYNC=true npm start`

### 1. `Name` property
1. Open wordpress.com and purchase a site with Business plan
2. Open Studio and connect domain
4. Open https://wordpress.com/settings/general/<DOMAIN>
5. Change `Title`
6. Open Studio -> "Sync" tab
7. Assert that the title is updated

### 2. Adding staging site
1. Open wordpress.com and purchase a site with Business plan
2. Open Studio and connect domain
3. Open wordprress.com and add staging site
8. Come back to Studio, reload
9. Assert that staging site was added to UI and appdata-v1.json:<br />Screenshot 2024-11-26 at 16 37 16<br />
Screenshot 2024-11-26 at 16 37 33

### 3. Removing staging site
1. We can test it following previous testing steps
2. So now open wordpress.com and remove staging site
3. Come back to Studio and assert that it's disappeared<br />Screenshot 2024-11-26 at 16 39 07<br />Screenshot 2024-11-26 at 16 39 25

### 4. Recreating staging site
1. Open wordpress.com and purchase a site with Business plan
4. Open `https://wordpress.com/staging-site/<DOMAIN_NAME>` and add staging site 
10. Open Studio and connect it
11. Assert that you see the connected site with staging<br />:Screenshot 2024-11-26 at 14 58 47
12. Assert that you see the same info in appdata-v1.json<br />:Screenshot 2024-11-26 at 14 59 28
13. Open wordpress.com again -> remove staging site -> and immediately create a new staging site. So our goal is to cover such case, when a site was recrerated, so we need to remove the old one from Studio and add a new one.
14. Open Studio and assert that you see the new URL<br />Screenshot 2024-11-26 at 15 01 53
15. Open appdata-v1.json and assert that you see the new `stagingSiteIds` and correctly removed the old one and added a new one staging site object. <br />Screenshot 2024-11-26 at 15 02 51

### 5. Try to test some edge cases with difficult scenarios, when you have two local sites and connect to them different / the same site, making changes in different order.

### CODE REVIEW

File: src/ipc-handlers.ts [hunk 1]
```diff
@@ -220,11 +220,14 @@ export async function updateSite(
 	return mergeSiteDetailsWithRunningDetails( userData.sites );
 }
 
+type DataConnectWpcomSite = { sites: SyncSite[]; localSiteId: string };
+
 export async function connectWpcomSite(
 	event: IpcMainInvokeEvent,
-	sites: SyncSite[],
-	localSiteId: string
+	data: DataConnectWpcomSite | DataConnectWpcomSite[]
```
Comment (wojtekn) [inline src/ipc-handlers.ts]: Would code be clearer and simpler if we always used the array?

File: src/ipc-handlers.ts [hunk 2]
```diff
@@ -220,11 +220,14 @@ export async function updateSite(
 	return mergeSiteDetailsWithRunningDetails( userData.sites );
 }
 
+type DataConnectWpcomSite = { sites: SyncSite[]; localSiteId: string };
```
Comment (wojtekn) [inline src/ipc-handlers.ts]: Would `WpcomSitesToConnect` be clearer?

File: src/ipc-handlers.ts [hunk 3]
```diff
@@ -238,48 +241,63 @@ export async function connectWpcomSite(
...
 		throw new Error( 'User not authenticated' );
 	}
 
-	const connections = userData.connectedWpcomSites?.[ currentUserId ] || [];
+	const connectedWpcomSites = userData.connectedWpcomSites;
 
-	const updatedConnections = connections.filter(
-		( conn ) => ! ( siteIds.includes( conn.id ) && conn.localSiteId === localSiteId )
-	);
+	// Totally unreal case, added it to make types below more clear and if this error happens, we definitely have something wrong
```
Comment (wojtekn) [inline src/ipc-handlers.ts]: Can we remove it if we are not convinced this could even happen? What if we handled such a case gracefully?

File: src/hooks/sync-sites/use-site-sync-management.ts [hunk 4]
```diff
@@ -4,28 +4,82 @@ import { useAuth } from '../use-auth';
...
 import { useSiteDetails } from '../use-site-details';
 
-const upToDateConnectedSites = (
+type UpToDateConnectedSitesReturn = {
+	updatedConnectedSites: SyncSite[];
+	toAdd: SyncSite[];
+	toDelete: { id: number; localSiteId: string }[];
+};
+
+export const upToDateConnectedSites = (
```
Comment (wojtekn) [inline src/hooks/sync-sites/use-site-sync-management.ts]: Could we name it like a function to clearly say what it does, e.g., `reconcileConnectedSites` or `getConnectedSitesUpdates`? Then, returned type `ReconciledConnectedSites` or `ConnectedSitesUpdates`?

**Comment by wojtekn** [comment]
I left some comments related to readability.

The feature works as expected.

File: src/ipc-handlers.ts (hunk 1 above)
Comment (nightnei) [reply src/ipc-handlers.ts]: I was woring about readability of suing this function, from my point of view is more readable when you do:
`connectWpcomSite( site )` instead of `connectWpcomSite( [ site ] )`. And to support it we need just one simple line of code `const list = Array.isArray( data ) ? data : [ data ];`

File: src/ipc-handlers.ts (hunk 2 above)
Comment (nightnei) [reply src/ipc-handlers.ts]: Answered the reason and my thoughts above

File: src/ipc-handlers.ts (hunk 3 above)
Comment (nightnei) [reply src/ipc-handlers.ts]: I added this comment and condition to make code below more obvious and avoid TS errors/warnings. Otherwise we need to do optional chaining and add extra conditions below, which are just confusing and unnecessary.
Initially, I tried to go that way, but it was not looking nice and even confused myself during development. Unfortunatelly. TS isn't clever enough, like in this case 😄

File: src/hooks/sync-sites/use-site-sync-management.ts (hunk 4 above)
Comment (nightnei) [reply src/hooks/sync-sites/use-site-sync-management.ts]: `reconcileConnectedSites` sounds good, I will update it 👍

**Review by fredrikekelund — APPROVED** [review]
Functionally, this works as expected 👍 I think we can do some work to improve readability, though. I opened #707 that branches off this PR with some suggestions, @nightnei. 

Also, to zoom out for a moment, I think it's fair to say that this is a fairly complex PR. We can do some stuff to help improve readability (as I tried in #707), but mainly, I think the complexity here is an indication that we should look into simplifying how data about connected sites is stored in `appdata-v1.json`.

I paired with @nightnei to get some help in understanding the background to this PR, and one suggestion we came up with was to store information about staging sites directly on the connected sites like so:

```json
{
  "id": 205290335,
  "localSiteId": "becdcc85-4ef7-4390-b182-29afae77fa64",
  "name": "f26d",
  "url": "https://f26d.dev",
  "isStaging": false,
  "stagingSites": {
    "235943942": "https://staging-8129-f26d.wpcomstaging.com"
  },
  "syncSupport": "already-connected"
}
```

That's definitely outside the scope of this PR, but if we weren't in the home stretch for the v1.3.0 release, I think it might have made sense to look into some other refactoring work here straight away.

File: src/hooks/sync-sites/use-site-sync-management.ts (hunk 4 above)
Comment (fredrikekelund) [reply src/hooks/sync-sites/use-site-sync-management.ts]: I agree that "reconciled" seems like a good fit here 👍

File: src/ipc-handlers.ts (hunk 1 above)
Comment (fredrikekelund) [reply src/ipc-handlers.ts]: I actually agree with @wojtekn that always passing either a single site or an array of sites makes more sense

File: src/ipc-handlers.ts [hunk 5]
```diff
@@ -238,48 +241,63 @@ export async function connectWpcomSite(
...
+	);
 }
 
+type DataDisconnectWpcomSite = { siteIds: number[]; localSiteId: string };
+
 export async function disconnectWpcomSite(
 	event: IpcMainInvokeEvent,
-	siteIds: number[],
-	localSiteId: string
+	data: DataDisconnectWpcomSite | DataDisconnectWpcomSite[]
```
Comment (fredrikekelund) [inline src/ipc-handlers.ts]: Could we not pass just `localSiteId` instead of also passing `siteIds`? Seems like unnecessary duplication

File: src/ipc-handlers.ts [hunk 6]
```diff
@@ -238,48 +241,63 @@ export async function connectWpcomSite(
...
 
-	const updatedConnections = connections.filter(
-		( conn ) => ! ( siteIds.includes( conn.id ) && conn.localSiteId === localSiteId )
-	);
+	// Totally unreal case, added it to make types below more clear and if this error happens, we definitely have something wrong
+	if ( ! connectedWpcomSites || ! connectedWpcomSites?.[ currentUserId ]?.length ) {
+		throw new Error(
+			'Something went wrong, since you are trying to disconnect something, but there are no stored connections yet'
+		);
+	}
```
Comment (fredrikekelund) [inline src/ipc-handlers.ts]: If `data` doesn't match any of the connected sites, we just silently fail. `connectedWpcomSites` being undefined could potentially indicate a more serious error (I'm not sure exactly what, though). However, if it's simply empty, I would be OK with also just silently failing.

File: src/hooks/use-fetch-wpcom-sites.tsx [hunk 7]
```diff
@@ -177,6 +181,6 @@ export const useFetchWpComSites = ( connectedSiteIds: number[] ) => {
 		syncSites,
 		isFetching: isFetchingSites.current,
 		isInitialized: isInitialized.current,
-		refetchSites: fetchSites,
+		refetchSites,
```
Comment (fredrikekelund) [inline src/hooks/use-fetch-wpcom-sites.tsx:184]: Why this change, @nightnei? Seems like the end result is pretty much identical to what we had previously

File: src/ipc-handlers.ts (hunk 1 above)
Comment (nightnei) [reply src/ipc-handlers.ts]: I don't have any concerns regarding it, so let's go with the suggestion, I will revert this code 👌

File: src/ipc-handlers.ts (hunk 5 above)
Comment (nightnei) [reply src/ipc-handlers.ts]: Hm, sorry, didn't get what do you mean, could you please give `type` for array of `arguments`, to understand your point

File: src/hooks/use-fetch-wpcom-sites.tsx (hunk 7 above)
Comment (nightnei) [reply src/hooks/use-fetch-wpcom-sites.tsx:184]: Yesterday, I resolved conflicts with master in the previous PR and removed this code, but it's actually is used [here](https://github.com/Automattic/studio/blob/trunk/src/hooks/sync-sites/use-listen-deep-link-connection.ts#L24).
Actually, it was better to create separate PR, maybe I will do it, if this one isn't merged till EOD.

File: src/ipc-handlers.ts (hunk 6 above)
Comment (nightnei) [reply src/ipc-handlers.ts]: The main point is helping TS to understand the code and for us to simplify maintenance in the future:
1) The first condition is not critical
2) But the second one (`.?` optional chaining) - make the return type of function as `SyncSite[] | undefined`, so I didn't want to have `undefined` as possible return type and adding conditions in places where this function is used.
Ofc, we can do something like this:
```
return (connectedWpcomSites || []).[ currentUserId ].filter( ( conn ) =>
		list.some( ( { localSiteId } ) => conn.localSiteId === localSiteId )
	);
```
But in this case, another developer (who will be working with this code in the future) could wonder - "how `connectedWpcomSites` can be `undefiend`". So it can confuse. But, thinking about it right now one more time, maybe `connectedWpcomSites || []` also good idea.
WDYT about my thoughts?

Screenshot 2024-11-29 at 13 31 33

File: src/ipc-handlers.ts (hunk 5 above)
Comment (fredrikekelund) [reply src/ipc-handlers.ts]: `localSiteId` is shared between the production site and staging site. Would it not suffice to use just that property to identify which site (production and staging) we want to disconnect?

File: src/ipc-handlers.ts (hunk 6 above)
Comment (fredrikekelund) [reply src/ipc-handlers.ts]:
```suggestion
	if ( ! Array.isArray( connectedWpcomSites?.[ currentUserId ] ) ) {
		throw new Error(
			'Something went wrong, since you are trying to disconnect something, but there are no stored connections yet'
		);
	}
```
Ultimately this is kind of nitpicky, so we don't need to dwell on it too much, but this suggestion seems to also account for your point about always returning an array.

File: src/ipc-handlers.ts (hunk 5 above)
Comment (nightnei) [reply src/ipc-handlers.ts]: It's not obvious at a glance, I spent some time to understand it :) So I will try to explaince you to give you more understand of `appdata-v1.json`.

Here is an example:
Screenshot 2024-11-29 at 14 43 37

1) `id` is id of a site from wp.com
2) `localSiteId` is id of site inside Studio

So we can connect the same `id` to different `localSiteId`s (the same wp.com site to 2+ local sites). As a result, when we a disconnecting some wp.com site from Studio site - we need to disconnect specific connection between wp.com site and local site.

File: src/ipc-handlers.ts (hunk 6 above)
Comment (nightnei) [reply src/ipc-handlers.ts]: I am ok with any approach to help TS understand what is going on inside this function 😄 I was looking for the best (between the worst) option, and the current one was looking good.
I checked your suggestion and TS is happy with it too, and as for me even simplier to read than mine, so I will apply it in the next commit, thanks 🙂

File: src/ipc-handlers.ts (hunk 5 above)
Comment (fredrikekelund) [reply src/ipc-handlers.ts]: Ah, my bad. Guess I was confused by the data structure again 🤦‍♂️

File: src/ipc-handlers.ts [hunk 8]
```diff
@@ -238,26 +236,31 @@ export async function connectWpcomSite(
...
 
-	return connections.filter( ( conn ) => conn.localSiteId === localSiteId );
+	return connections.filter( ( conn ) =>
+		list.some( ( { localSiteId } ) => conn.localSiteId === localSiteId )
+	);
 }
 
+type WpcomSitesToDisconnect = { siteIds: number[]; localSiteId: string }[];
+
 export async function disconnectWpcomSite(
```
Comment (fredrikekelund) [inline src/ipc-handlers.ts]:
```suggestion
export async function disconnectWpcomSites(
```
Nit, but since this function takes an array of IDs, maybe we should rename it?

File: src/ipc-handlers.ts [hunk 9]
```diff
@@ -220,11 +220,9 @@ export async function updateSite(
...
 }
 
-export async function connectWpcomSite(
-	event: IpcMainInvokeEvent,
-	sites: SyncSite[],
-	localSiteId: string
-) {
+type WpcomSitesToConnect = { sites: SyncSite[]; localSiteId: string }[];
+
+export async function connectWpcomSite( event: IpcMainInvokeEvent, list: WpcomSitesToConnect ) {
```
Comment (fredrikekelund) [inline src/ipc-handlers.ts]:
```suggestion
export async function connectWpcomSites( event: IpcMainInvokeEvent, list: WpcomSitesToConnect ) {
```
Same thing here about potentially renaming.

File: src/ipc-handlers.ts (hunk 5 above)
Comment (nightnei) [reply src/ipc-handlers.ts]: Initially, I was also confused many times while working around it
So, it's totally ok :)

File: src/hooks/tests/up-to-date-connected-sites.test.ts [hunk 10]
```diff
@@ -0,0 +1,380 @@
+import { getUpdatedConnectedSites } from '../sync-sites/use-site-sync-management';
```
Comment (wojtekn) [inline src/hooks/tests/up-to-date-connected-sites.test.ts]: Can we just fix this import and rename tests to account for latest changes?
